lib/SILOptimizer/Transforms/StringSwitch.cpp

```diff
@@ -0,0 +1,258 @@
+//===--- StringSwitch.cpp - Optimization of string comparisons ------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//===----------------------------------------------------------------------===//
+
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SILOptimizer/Analysis/AliasAnalysis.h"
+#include "swift/SILOptimizer/Analysis/DominanceAnalysis.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/ConstExpr.h"
+
+using namespace swift;
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+//                            StringSwitchPass
+//===----------------------------------------------------------------------===//
+
+// Optimize multiple string literal comparisons against a single String type.
+// This can be used to optimize a series of string comparison ifs
+// or, more commonly, a string-switch-statement.
+class StringSwitchPass : public SILFunctionTransform {
+  bool doOneIteration() {
+    SILFunction *F = getFunction();
+    SILModule &module = F->getModule();
+
+    DominanceAnalysis *domAnalysis = PM->getAnalysis<DominanceAnalysis>();
+    DominanceInfo *domInfo = domAnalysis->get(F);
+    DominanceOrder domOrder(&F->front(), domInfo);
+
+    // Find all the string literals used in a series of comparisons agains
+    // one "normal" string.
+    SmallVector<StringLiteralInst *, 12> stringLiterals;
+    SmallVector<ApplyInst *, 12> stringCompares;
+    SILValue subject;
+    // Iterate through all the blocks in the function using dominance order
+    // so that we don't miss the first comparison because we started with the
+    // seconcd comparison (hypothetically).
+    while (auto *block = domOrder.getNext()) {
```

Comment:
Sorry that I didn't notice this earlier, but this is quadratic complexity in worst case. Imagine a large function with _a lot of_ string compares. This algorithm would traverse the whole function for each string compare.

Is it possible to handle the whole function with a single traverse of the dominator tree? You could use a data structure like a DenseMap<SILValue, SmallVector<ApplyInst>> which maps subjects to the list of string compares.

Reply:
I'm pretty sure this is linear (minus the dominance order stuff). I traverse each function once (`doOneIteration` is called once for every function). 

If the optimization succeeds and modifies the function, then it will run again, but I don't think that's what you're talking about.

Reply:
In run() there is
```
   while (doOneIteration()) {
       changed = true;
     }
```

If doOneIteration converts one string comparison at a time (and if there are many unrelated comparison in the function), then this is quadratic.

Reply:
Ahh, I see what you mean. I'll refactor it a bit and fix this.

Reply:
I've refactored `StringSwitch` a bit. Now all the helper variables (mostly types) are members so they aren't initialized multiple times. I also moved the collection and optimization into two different functions. Now it only iterates through each function once (twice if it finds an optimization).

Reply:
Locals are better. Member variables in a function pass are kind of "global" variables, which are usually hard to figure out where they are set and read.

Reply:
Fair enough. Will do. Thanks.

Reply:
My bad, because `optimizeStringCompare` is called (possibly) many times, I think it would still probably be faster to just initialize them once (instead of every time we optimize a string comparison). Should I still make them locals?